File: tools/recdefgen.rb
#!/usr/bin/env ruby
# PANDA -- a simple transaction monitor
# Copyright (C) 2005-2008  ogochan
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA
require "xmlparser"
require "kconv"
include Kconv
require "uconv"
include Uconv

$button_enable = false   
if ARGV[0] == "-b" 
  $button_enable = true 
  ARGV.shift 
end     

def  putTab(n)
  while  n > 0
    printf("\t");
    n -= 1;
  end
end

class  Stack
  def  initialize
    @st = Array.new;
  end
  def  push(val)
    @st.push(val);
  end
  def  pop
    @st.pop;
  end
  def  top
    @st[@st.size-1];
  end
  def  body
    @st;
  end
end

class  Ssignal

  attr_accessor :name,:handler,:data

  def  initizlize
    @handler = "";
    @name = "";
    @data = "";
  end
end

class  Widget

  attr_accessor :child,:child_name,:chars,:klass,:name,:label,:format,
    :width,:height,:columns,:rows,:column_width,:align,:signal,:group

  def  initialize
    @child = Array.new;
    @child_name = "";
    @chars = 256;
    @klass = "";
    @name = "";
    @label = "";
    @format = "";
    @width = 0;
    @height = 0;
    @columns = 0;
    @rows = 0;
    @column_width = Array.new;
    @align = "left";
    @signal = Array.new;
    @group = "";
  end

  def  _windowName
    case  @klass
      when  "GtkWindow"
      title = @name;
      else
      title = "";
      for  c in @child
        title = c._windowName;
        break  if  title  !=  "";
      end
    end
    return  (title);
  end
  def  _title
    case  @klass
      when  "GtkWindow"
      title = @label;
      else
      title = "";
      for  c in @child
        title = c._title;
        break  if  title  !=  "";
      end
    end
    return  (title);
  end
  def  isData
    ret = FALSE
    case  @klass
      when  "GtkEntry", "GtkToggleButton", "GtkCheckButton", 
        "GtkRadioButton", "GtkList", "GtkCalendar", 
        "GtkNumberEntry", "GtkText", "GtkPandaEntry", 
        "GtkPandaText" ,"GtkCList", "GtkPandaCList", 
        "GtkPandaPS", "GnomePixmap", "GnomeFileEntry",
		"GtkFileChooserButton","GtkColorButton",
        "GtkPandaPDF", "GtkPandaDownload", "GtkPandaPrint","GtkPandaTable",
        "GtkWindow","GtkNotebook"
        ret = TRUE;
      when  "GtkLabel"
        if  @label == "" or
        @label =~ /^X+$/
        ret = TRUE;
      end
      else
      if  @child
        for  c in @child
          ret = c.isData;
          break  if  ret;
        end
        else
        ret = FALSE;
      end
    end
    return  (ret);
  end
  def  _panda(ind)
    vname = @name;
    size = 0;
    case  @klass
      when  "top-level"
      for  c in @child
        c._panda(ind);
      end
      when  "GtkVBox", "GtkHBox", "GtkTable", "GtkFixed", 
        "GtkScrolledWindow", "GtkViewport", "GtkNotebook", 
        "GtkFrame", "GtkHandleBox", "GtkMenuBar", "GtkToolbar"
        if  self.isData
          putTab(ind);
          printf("%s\t{\n",vname);
          if    @klass  ==  "GtkNotebook"
            putTab(ind+1);
            printf("pageno\tint;\n");
          end
          if    @klass  ==  "GtkScrolledWindow"
            putTab(ind+1); 
            printf("hpos\tint;\n"); 
            putTab(ind+1); 
            printf("vpos\tint;\n"); 
          end
          for  c in @child
            c._panda(ind+1);
          end
          putTab(ind);
          printf("};\n");
        end
      when "GtkWindow"
        if  self.isData
          putTab(ind);
          printf("%s\t{\n",vname);

        	putTab(ind+1);
        	printf("title\tvarchar(256);\n");
        	putTab(ind+1);
        	printf("popup_summary\tvarchar(256);\n");
        	putTab(ind+1);
        	printf("popup_body\tvarchar(1024);\n");
        	putTab(ind+1);
        	printf("popup_icon\tvarchar(256);\n");
        	putTab(ind+1);
        	printf("popup_timeout\tint;\n");

          for  c in @child
            c._panda(ind+1);
          end
          putTab(ind);
          printf("};\n");
        end
      when  "GtkCList", "GtkPandaCList"
        putTab(ind);
        printf("%s\t{\n",vname);
        i = 0;
        for  c in @child
          c._panda(ind+1);
          i += 1;
        end

        putTab(ind+1);
        printf("count\tint;\n");
        putTab(ind+1);
        if @klass == "GtkPandaCList"
          printf("row\tint;\n");
          putTab(ind+1);
        end
        printf("item\t{\n");
        i = 0;
        for  c in @child
          putTab(ind+2);
          printf("value%d\tvarchar(%d);\n",i,Integer(self.column_width[i])/8);
          i += 1;
        end
        putTab(ind+1);
        printf("}\t[??];\n");
        putTab(ind+1);
        printf("select\tbool[??];\n");

        putTab(ind);
        printf("};\n");
      when  "GtkList"
        putTab(ind);
        printf("%s\t{\n",vname);
        i = 0;
        for  c in @child
          c._panda(ind+1);
          i += 1;
        end
        putTab(ind+1);
        printf("count\tint;\n");
        putTab(ind+1);
        printf("item\tvarchar(??)[??];\n");
        putTab(ind+1);
        printf("select\tbool[??];\n");

        putTab(ind);
        printf("};\n");
      when  "GtkOptionMenu"
        putTab(ind);
        printf("%s\t{\n",vname);
        i = 0;
        for  c in @child
          c._panda(ind+1);
          i += 1;
        end
        putTab(ind+1);
        printf("count\tint;\n");
        putTab(ind+1);
        printf("item\tvarchar(??)[??];\n");
        putTab(ind+1);
        printf("select\tint;\n");
        putTab(ind);
        printf("};\n");
      when  "GtkCombo", "GtkPandaCombo"
        putTab(ind);
        printf("%s\t{\n",vname);
        @child[0]._panda(ind+1);
        putTab(ind+1);
        printf("count\tint;\n");
        putTab(ind+1);
        printf("item\tvarchar(%d)[??];\n",@child[0].chars);
        putTab(ind);
        printf("};\n");
      when  "GtkLabel"
      if    @label  ==  "" or
          @label  =~  /^X+$/
        putTab(ind);
        printf("%s\t{\n",vname);
        putTab(ind+1);
        if  @label  ==  ""
          printf("value\tvarchar(%d);\n",Integer(self.width)/8);
          else
          printf("value\tvarchar(%d);\n",label.length);
        end
        putTab(ind);
        printf("};\n");
      end
      when  "GtkEntry", "GtkPandaEntry"
        putTab(ind);
        printf("%s\t{\n",vname);
        putTab(ind+1);
        printf("value\tvarchar(%d);\n",@chars);
        putTab(ind);
        printf("};\n");
      when  "GtkText", "GtkPandaText"
        putTab(ind);
        printf("%s\t{\n",vname);
        putTab(ind+1);
        printf("value\tvarchar(??);\n");
        putTab(ind);
        printf("};\n");
      when  "GtkNumberEntry"
        putTab(ind);
        printf("%s\t{\n",vname);
        putTab(ind+1);
        len = @format.tr(".,","").length;
        s = @format.split(".");
        if    s[1]  !=  nil
          slen = s[1].tr(",","").length;
          printf("value\tnumber(%d,%d);\n",len,slen);
          else
          printf("value\tnumber(%d);\n",len);
        end;
        putTab(ind);
        printf("};\n");
      when  "GtkToggleButton", "GtkCheckButton", "GtkRadioButton"
        putTab(ind);
        printf("%s\t{\n",vname);
        putTab(ind+1);
        printf("value\tbool;\n");
        if    @label  ==  ""
          putTab(ind+1);
          printf("label\tvarchar(??);\n");
        end
        putTab(ind);
        printf("};\n");
      when  "GtkButton" 
      if  $button_enable 
        putTab(ind); 
        printf("%s\t{\n",vname); 
        putTab(ind+1); 
        printf("state int;\n"); 
        putTab(ind); 
        printf("};\n"); 
      end 
      when  "GtkProgressBar" 
        putTab(ind); 
        printf("%s\t{\n",vname); 
        putTab(ind+1); 
        printf("value\tint;\n"); 
        putTab(ind); 
        printf("};\n"); 
      when  "GtkCalendar"
        putTab(ind);
        printf("%s\t{\n",vname);
        putTab(ind+1);
        printf("year\tint;\n");
        putTab(ind+1);
        printf("month\tint;\n");
        putTab(ind+1);
        printf("day\tint;\n");
        putTab(ind);
        printf("};\n");
      when  "GnomePixmap", "GtkPandaPS", "GtkPandaPDF"
        putTab(ind);
        printf("%s\t{\n",vname);
        putTab(ind+1);
          printf("objectdata\tobject;\n");
        putTab(ind);
        printf("};\n");
      when  "GtkPandaDownload"
        putTab(ind);
        printf("%s\t{\n",vname);
          putTab(ind+1);
          printf("objectdata\tobject;\n");
          putTab(ind+1);
          printf("filename\tvarchar(%d);\n",@chars);
          putTab(ind+1);
          printf("description\tvarchar(%d);\n",@chars);
        putTab(ind);
        printf("};\n");
      when  "GtkPandaPrint"
        putTab(ind);
        printf("%s\t{\n",vname);
          putTab(ind+1);
          printf("item {\n");
            putTab(ind+2);
            printf("url\tvarchar(%d);\n",@chars);
            putTab(ind+2);
            printf("title\tvarchar(%d);\n",@chars);
            putTab(ind+2);
            printf("nretry\tint;\n");
            putTab(ind+2);
            printf("showdialog\tint;\n");
          putTab(ind+1);
          printf("}[10];\n");
        putTab(ind);
        printf("};\n");
      when  "GnomeFileEntry"
        putTab(ind);
        printf("%s\t{\n",vname);
          putTab(ind+1);
          printf("objectdata\tobject;\n");
            putTab(ind+1);
            printf("value\tvarchar(%d);\n",@chars);
        putTab(ind);
        printf("};\n");
      when  "GtkFileChooserButton"
        putTab(ind);
        printf("%s\t{\n",vname);
          putTab(ind+1);
          printf("objectdata\tobject;\n");
          putTab(ind+1);
          printf("filename\tvarchar(%d);\n",@chars);
        putTab(ind);
        printf("};\n");
      when  "GtkColorButton"
        putTab(ind);
        printf("%s\t{\n",vname);
          putTab(ind+1);
          printf("color\tvarchar(%d);\n",@chars);
        putTab(ind);
        printf("};\n");
      when  "GtkPandaTable"
        putTab(ind);
        printf("%s\t{\n",vname);

        putTab(ind+1);
        printf("state\tint;\n");
        putTab(ind+1);
        printf("style\tvarchar(%d);\n",@chars);
        putTab(ind+1);
        printf("trow\tint;\n");
        putTab(ind+1);
        printf("trowattr\tint;\n");
        putTab(ind+1);
        printf("tcolumn\tint;\n");
        putTab(ind+1);
        printf("tvalue\tvarchar(%d);\n",@chars);
        putTab(ind+1);
        printf("rowdata {\n");
        @columns.times{|i|
          putTab(ind+2);
          printf("column%d {\n",i+1);
            putTab(ind+3);
            printf("celldata\tvarchar(%d);\n",@chars);
            putTab(ind+3);
            printf("fgcolor\tvarchar(%d);\n",@chars);
            putTab(ind+3);
            printf("bgcolor\tvarchar(%d);\n",@chars);
          putTab(ind+2);
          printf("};\n");
        }
        putTab(ind+1);
        printf("}[n_rows];\n");

        putTab(ind);
        printf("};\n");
      else
      ;
    end
  end
  def  panda
    self._panda(0);
  end
end

if ((xml = $<.gets).nil?); exit 1; end

if xml =~ /^<\?xml\sversion=.+\sencoding=.EUC-JP./i
  xml.sub!(/EUC-JP/i, "UTF-8")
  encoding = 'EUC-JP'
elsif xml =~ /^<\?xml\sversion=.+\sencoding=.Shift_JIS./i
  xml.sub!(/Shift_JIS/i, "UTF-8")
  encoding = "Shift_JIS"
elsif xml =~ /^<\?xml\sversion=.+\sencoding=.ISO-2022-JP./i
  xml.sub!(/ISO-2022-JP/i, "UTF-8")
  encoding = "ISO-2022-JP"
elsif xml =~ /^<\?xml\sversion=.+\sencoding=.UTF-8./i
  encoding = "UTF-8"
else
  xml.sub!(/EUC-JP/i, "UTF-8")
  encoding = 'EUC-JP'
end

xml += String($<.read)

if encoding == "EUC-JP"
  xml = euctou8(xml)
elsif encoding == "Shift_JIS"
  xml = euctou8(kconv(xml, EUC, SJIS))
elsif encoding == "ISO-2022-JP"
  xml = euctou8(kconv(xml, EUC, JIS))
end

parser = XMLParser.new

tag = Stack.new;
tree = Stack.new;

widget = Widget.new;
tree.push(widget);
widget.name = "top";
widget.klass = "top-level";

signal = nil;

begin
  current = 0;
  parser.parse(xml) do  |type, name, data|
    case  type
      when  XMLParser::START_ELEM
      tag.push(name);
      case  name
        when  /widget/i
        widget = Widget.new;
        tree.push(widget);
        when  /signal/i
        signal = Ssignal.new;
        else
        ;
      end
      when  XMLParser::END_ELEM
      tag.pop;
      case  name
        when  /widget/i
        cur = tree.pop;
        widget = tree.top;
        widget.child << cur;
        when  /signal/i
        widget.signal << signal;
        else
        ;
      end
      when  XMLParser::CDATA
      next if data =~ /^\s*$/;
      data = Uconv.u8toeuc(data)
      path = "";
      for  t in tag.body
        path = path + "/" + t;
      end
      case  path
        when  /\/class$/i
        widget.klass = data;
        when  /\/widget\/name$/i
        widget.name = data;
        when  /\/text_max_length$/i
        widget.chars = Integer(data);
        when  /\/max_length$/i
        widget.chars = Integer(data);
        when  /\/widget\/label$/i
        widget.label = data;
        when  /\/widget\/title/i
        widget.label = data;
        when  /\/widget\/width/i
        widget.width = Integer(data);
        when  /\/widget\/height/i
        widget.height = Integer(data);
        when  /\/widget\/columns/i
        widget.columns = Integer(data);
        when  /\/widget\/rows/i
        widget.rows = Integer(data);
        when  /\/widget\/format/i
        widget.format = data;
        when  /\/widget\/group/i
        widget.group = data;
        when  /\/widget\/child_name/i
        widget.child_name = data;
        when  /\/widget\/column_width/i
        widget.column_width = data.split(",");

        when  /\/widget\/justify/i
        case  data
          when  /GTK_JUSTIFY_CENTER/i
          widget.align = "center";
          when  /GTK_JUSTIFY_LEFT/i
          widget.align = "left";
          when  /GTK_JUSTIFY_RIGHT/i
          widget.align = "right";
          else
          ;
        end
        when  /\/signal\/name/i
        signal.name = data;
        when  /\/signal\/handler/i
        signal.handler = data;
        when  /\/signal\/data/i
        signal.data = data;
        else
        ;
      end
      when  XMLParser::PI
      else
      next if current == 0
      data.gsub!("\n", "\\n")
      printf("%s\n",Uconv.u8toeuc(data));
      current = 0
    end
  end
rescue XMLParserError
  line = parser.line
  column = parser.column
  printf("%s:%s:%d:%d:E:%s\n",$0,$<.filename,parser.line,parser.column,$!);
  exit 1
end

widget.panda;
exit 0

File: tools/recdefgen.c
#if 0
#!/bin/bash
src=$0
obj=${src%.*}
gcc -g -Wl,--no-as-needed `pkg-config --cflags --libs libglade-panda-2.0` -o $obj $src
$obj
exit
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <gtk/gtk.h>
#include <gtkpanda/gtkpanda.h>
#include <glade/glade.h>

#define DEFAULT_CHAR_SIZE 256
#define DEFAULT_ARRAY_SIZE 64

void
putLevel(int l)
{
  int i;
  for(i=0;i<l;i++) {
    printf("  ");
  }
}

void
GenRecord(
  GtkWidget *widget,
  int level)
{
  long type;
  int i,length,len,slen;
  char *format,*p;
  gboolean haveDot;

  putLevel(level);
  printf("%s { #%s\n",
   glade_get_widget_name(widget),G_OBJECT_TYPE_NAME(widget));

  type = (long)(G_OBJECT_TYPE(widget));
  if (type == GTK_TYPE_NOTEBOOK) {
    putLevel(level+1);
    printf("pageno int;\n");
  } else if (type == GTK_TYPE_WINDOW) {
    putLevel(level+1);
    printf("title varchar(%d);\n",DEFAULT_CHAR_SIZE);
    putLevel(level+1);
    printf("popup_summary varchar(%d);\n",DEFAULT_CHAR_SIZE);
    putLevel(level+1);
    printf("popup_body varchar(%d);\n",DEFAULT_CHAR_SIZE);
    putLevel(level+1);
    printf("popup_icon varchar(%d);\n",DEFAULT_CHAR_SIZE);
    putLevel(level+1);
    printf("popup_timeout int;\n",DEFAULT_CHAR_SIZE);
  } else if (type == GTK_PANDA_TYPE_CLIST) {
    putLevel(level+1);
    printf("count int;\n");
    putLevel(level+1);
    printf("row int;\n");
    putLevel(level+1);
    printf("selectdata bool[%d];\n",DEFAULT_ARRAY_SIZE);
    putLevel(level+1);
    printf("fgcolor varchar(%d)[%d];\n",DEFAULT_CHAR_SIZE,DEFAULT_ARRAY_SIZE);
    putLevel(level+1);
    printf("bgcolor varchar(%d)[%d];\n",DEFAULT_CHAR_SIZE,DEFAULT_ARRAY_SIZE);
    putLevel(level+1);
    printf("item {\n");
    for(i=0;i<gtk_panda_clist_get_columns(GTK_PANDA_CLIST(widget));i++) {
      putLevel(level+2);
      printf("column%d varchar(%d);\n",i+1,DEFAULT_CHAR_SIZE);
    }
    putLevel(level+1);
    printf("}[%d];\n",DEFAULT_ARRAY_SIZE);
  } else if (type == GTK_PANDA_TYPE_TABLE) {
    putLevel(level+1);
    printf("state int;\n");
    putLevel(level+1);
    printf("style varchar(%d);\n",DEFAULT_CHAR_SIZE);
    putLevel(level+1);
    printf("trow int;\n");
    putLevel(level+1);
    printf("trowattr int;\n");
    putLevel(level+1);
    printf("tcolumn int;\n");
    putLevel(level+1);
    printf("tvalue varchar(%d);\n",DEFAULT_CHAR_SIZE);
    putLevel(level+1);
    printf("rowdata {\n");
    for(i=0;i<gtk_panda_table_get_columns(GTK_PANDA_TABLE(widget));i++) {
      putLevel(level+2);
      printf("column%d {\n",i+1);
        putLevel(level+3);
        printf("celldata varchar(%d);\n",DEFAULT_CHAR_SIZE);
        putLevel(level+3);
        printf("fgcolor varchar(%d);\n",DEFAULT_CHAR_SIZE);
        putLevel(level+3);
        printf("bgcolor varchar(%d);\n",DEFAULT_CHAR_SIZE);
      putLevel(level+2);
      printf("};\n\n",i+1);
    }
    putLevel(level+1);
    printf("};\n");
  } else if (type == GTK_PANDA_TYPE_COMBO) {
    putLevel(level+1);
    printf("count int;\n");
    putLevel(level+1);
    printf("item varchar(%d)[%d];\n",DEFAULT_CHAR_SIZE,DEFAULT_ARRAY_SIZE);
  } else if (type == GTK_TYPE_LABEL) {
    putLevel(level+1);
    printf("textdata varchar(%d);\n",DEFAULT_CHAR_SIZE);
  } else if (type == GTK_PANDA_TYPE_ENTRY) {
    length = gtk_entry_get_max_length(GTK_ENTRY(widget));
    if (length <= 0) {
      length = DEFAULT_CHAR_SIZE;
    }
    putLevel(level+1);
    printf("textdata varchar(%d);\n",length);
  } else if (type == GTK_PANDA_TYPE_TEXT) {
    putLevel(level+1);
    printf("textdata varchar(%d);\n",DEFAULT_CHAR_SIZE);
  } else if (type == GTK_TYPE_NUMBER_ENTRY) {
    g_object_get(G_OBJECT(widget),"format",&format,NULL);
    len = slen = 0;
    haveDot = FALSE;
    for(p = format; (p != NULL && *p != 0); p++) {
      if (*p == ',') {
        continue;
      }
      if (*p == '.') {
        haveDot = TRUE;
        continue;
      }
      len++;
      if (haveDot) {
        slen++;
      }
    }
    putLevel(level+1);
    if (slen != 0) {
      printf("numdata number(%d,%d);\n",len,slen);
    } else {
      printf("numdata number(%d);\n",len);
    }
  } else if (
    type == GTK_TYPE_TOGGLE_BUTTON ||
    type == GTK_TYPE_CHECK_BUTTON ||
    type == GTK_TYPE_RADIO_BUTTON
  ) {
    putLevel(level+1);
    printf("isactive bool;\n");
    putLevel(level+1);
    printf("label varchar(%d);\n",DEFAULT_CHAR_SIZE);
  } else if (type == GTK_TYPE_BUTTON) {
    putLevel(level+1);
    printf("state int;\n");
  } else if (type == GTK_TYPE_PROGRESS_BAR) {
    putLevel(level+1);
    printf("value int;\n");
  } else if (type == GTK_TYPE_CALENDAR) {
    putLevel(level+1);
    printf("year int;\n");
    putLevel(level+1);
    printf("month int;\n");
    putLevel(level+1);
    printf("day int;\n");
  } else if (
    type == GTK_PANDA_TYPE_PDF ||
    type == GTK_PANDA_TYPE_PIXMAP
  ) {
    putLevel(level+1);
    printf("objectdata object;\n");
  } else if (type == GTK_PANDA_TYPE_DOWNLOAD) {
    putLevel(level+1);
    printf("objectdata object;\n");
    putLevel(level+1);
    printf("filename varchar(%d);\n",DEFAULT_CHAR_SIZE);
    putLevel(level+1);
    printf("description varchar(%d);\n",DEFAULT_CHAR_SIZE);
  } else if (type == GTK_PANDA_TYPE_DOWNLOAD2) {
    putLevel(level+1);
    printf("item {\n");
      putLevel(level+2);
      printf("path varchar(%d)\n",DEFAULT_CHAR_SIZE);
      putLevel(level+2);
      printf("filename varchar(%d)\n",DEFAULT_CHAR_SIZE);
      putLevel(level+2);
      printf("description varchar(%d)\n",DEFAULT_CHAR_SIZE);
      putLevel(level+2);
      printf("nretry int\n");
    putLevel(level+1);
    printf("}[%d];\n",DEFAULT_ARRAY_SIZE);
  } else if (type == GTK_PANDA_TYPE_PRINT) {
    putLevel(level+1);
    printf("item {\n");
      putLevel(level+2);
      printf("path varchar(%d)\n",DEFAULT_CHAR_SIZE);
      putLevel(level+2);
      printf("title varchar(%d)\n",DEFAULT_CHAR_SIZE);
      putLevel(level+2);
      printf("nretry int\n");
      putLevel(level+2);
      printf("showdialog int\n");
    putLevel(level+1);
    printf("}[%d];\n",DEFAULT_ARRAY_SIZE);
  } else if (type == GTK_TYPE_FILE_CHOOSER_BUTTON) {
    putLevel(level+1);
    printf("objectdata object;\n");
    putLevel(level+1);
    printf("filename varchar(%d);\n",DEFAULT_CHAR_SIZE);
  } else if (type == GTK_TYPE_COLOR_BUTTON) {
    putLevel(level+1);
    printf("color varchar(%d);\n",DEFAULT_CHAR_SIZE);
  }
}

void
GenRecordEnd(
  int l)
{
  putLevel(l);
  printf("};\n");
}

void
GenRecordDefine(
  GtkWidget *widget,
  gpointer data)
{
  GtkWidget *child;
  int level;
  long type;

  level = GPOINTER_TO_INT(data);
  type = (long)(G_OBJECT_TYPE(widget));

  if (
    type == GTK_TYPE_HSEPARATOR||
    type == GTK_TYPE_VSEPARATOR
  ) {
    return;
  }

  if (!glade_get_widget_name(widget)) {
    return;
  }

  if (GTK_IS_CONTAINER(widget)) {
    GenRecord(widget,level);

    gtk_container_forall(GTK_CONTAINER(widget),
      GenRecordDefine,
      GINT_TO_POINTER(level+1));

    if (GTK_IS_WINDOW(widget)) {
      child = (GtkWidget *)g_object_get_data(G_OBJECT(widget), "child");
      if (child) {
        GenRecordDefine(child,GINT_TO_POINTER(level+1));
      }
    }

    GenRecordEnd(level);
  } else {
    GenRecord(widget,level);
    GenRecordEnd(level);
  }
}

int
main(int argc,char *argv[])
{
  GladeXML *xml;
  GtkWidget *root;
  GRegex *re;
  GMatchInfo *m;
  char *win;

  if (argc < 2) {
    fprintf(stderr,"./recdefgen window.glade\n");
    exit(-1);
  }

  win = NULL;
  re = g_regex_new("\\/([a-zA-Z0-9_-]+).glade",G_REGEX_CASELESS,0,NULL);
  if (g_regex_match(re,argv[1],0,&m)) {
    win = g_match_info_fetch(m,1);
    g_match_info_free(m);
  }
  g_regex_unref(re);

  gtk_init(&argc,&argv);
  gtk_panda_init(&argc,&argv);
  glade_init();
  xml = glade_xml_new(argv[1],win);

  if (!xml) {
    g_error("glade_xml_new error:%s",argv[1]);
  }

  root = glade_xml_get_widget(xml,win); 

  if (!root) {
    g_error("cant get root widget:%s",win);
  }

  GenRecordDefine(root,GINT_TO_POINTER(0));  

  return 0;
}
